local-cli/server/util/attachHMRServer.js

```diff
@@ -16,19 +16,262 @@ const url = require('url');
  * Hot Module Replacement updates to the simulator.
  */
 function attachHMRServer({httpServer, path, packagerServer}) {
-  let client = null;
 
-  function disconnect() {
-    client = null;
-    packagerServer.setHMRFileChangeListener(null);
+  let clients = [];
+  let clientsPerBundleEntry = {};
+
+  let platformCache = {
+    android: {},
+    ios: {},
+  };
+
+  function addClient (client) {
+    client.ws.on('error', e => {
+      console.error('[Hot Module Replacement] Unexpected error', e);
+      disconnect(client);
+    });
+
+    client.ws.on('close', () => disconnect(client));
+    clients.push(client);
+
+    clientsPerBundleEntry[client.bundleEntry] = [...(
+      clientsPerBundleEntry[client.bundleEntry] ? clientsPerBundleEntry[client.bundleEntry] : []
+    ), client];
+
+    // Only register callback if it's the first client
+    if (clients.length === 1) {
+      packagerServer.setHMRFileChangeListener(onHMRChange);
+    }
+
+  }
+
+  function disconnect (disconnectedClient) {
+    clients = clients.filter(client => client !== disconnectedClient);
+    clientsPerBundleEntry[disconnectedClient.bundleEntry] = clientsPerBundleEntry[disconnectedClient.bundleEntry].filter(client => client !== disconnectedClient);
+    //Only clear change listener if there are no more listening clients
+    if (clients.length === 0) {
+      packagerServer.setHMRFileChangeListener(null);
+    }
+  }
+
+  function connectedPlatformClients () {
+    return {
+      android: clients.filter(client => client.platform === "android"),
+      ios: clients.filter(client => client.platform === "ios"),
+    }
+  }
+
+  function forEachPlatform (callback) {
+    const platformClients = connectedPlatformClients();
+    return Promise.all([
+      wrapPotentialPromise(platformClients.android.length > 0 ? callback("android", platformClients.android) : null),
+      wrapPotentialPromise(platformClients.ios.length > 0 ? callback("ios", platformClients.ios) : null),
+    ]).then(results => ({
+       android: results[0],
+       ios: results[1],
+    }));
+  }
+
+  function forEachBundleEntry (callback) {
+    return forEachPlatform(platform =>
+      forEachKey(platformCache[platform], bundleEntry =>
+        callback(bundleEntry, platform)
+      )
+    )
+  }
+
+  function sendToClients (message, platform, bundleEntry) {
+    if (platform && bundleEntry) {
+      clientsPerBundleEntry[bundleEntry].map(client => {
+        if (client.platform === platform) {
+          client.ws.send(JSON.stringify(message))
```

Comment:
I modified a similar feature for myself, sometimes `send` raises exceptions, so need catch it. Maybe `Broken Pipe`, I don't remember.